modules/telemetry/include/hephaestus/telemetry/struclog.h

```diff
@@ -0,0 +1,145 @@
+#pragma once
+
+#include <iomanip>
+#include <source_location>
+#include <sstream>
+#include <string>
+#include <string_view>
+#include <thread>
+#include <vector>
+
+namespace heph::telemetry {
+
+enum class Level : std::uint8_t { Trace, Debug, Info, Warn, Error, Fatal };
+auto operator<<(std::ostream& /*os*/, const Level& /*level*/) -> std::ostream&;
+
+template <typename T>
+concept NotQuotable = !std::convertible_to<T, std::string> && requires(std::ostream& os, T t) {
+  { os << t } -> std::same_as<std::ostream&>;
+};
+
+template <typename T>
+concept Quotable = std::convertible_to<T, std::string>;
+
+template <typename T>
+struct Field final {
+  std::string key;
+  T val;
+};
+
+/**
+ * @brief A class that allows easy composition of logs for structured logging.
+ *        Example(see also struclog.cpp):
+ *        namespace ht=heph::telemetry;
+ *        log(ht::LogEntry{Level::Info, "adding"} | "id"_f(12345) | ht::Field{"tag", "test"}) ;
+ *
+ *        logs
+ *        'level=info file=struclog.h:123 time=2023-12-3T8:52:02+0 message="adding" id=12345 tag="test"'
+ *
+ *        Remark: We would like to use libfmt with quoted formatting as proposed in
+ *                https://github.com/fmtlib/fmt/issues/825 once its included (instead of sstream).
+ *
+ */
+class LogEntry {
+public:
+  using Fields = std::vector<Field<std::string>>;
```

Comment:
I would have like for metric a variant here, actually we can use the same variant.

Reply:
Finally coming back to this :) To clarify: you would prefer saving the typed values in a variant instead of converting them to string and saving them in the LogEntry? the motivation behind this was to distinguish types that need special rendering in the | operator throught concepts (here Quotable and NonQuotable) and save them all in a format that can be piped to the output sink. I thought that it would anyway be streamed to some string format so rendering them there would remove some resp from the sink implementor...

Reply:
Valid point. In the case of metrics we want variants because the sink will also need to use them as numbers for e.g. influxdb. But in this case you are right, a string should be sufficient as otherwise it should probably be metric.